gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/InsertAllRequest.java

```diff
@@ -0,0 +1,399 @@
+ * choosing an approach.
+ *
+ * @see <a href="https://cloud.google.com/bigquery/streaming-data-into-bigquery">Streaming Data into
+ *     BigQuery</a>
+ */
+public class InsertAllRequest implements Serializable {
+
+  private static final long serialVersionUID = 211200307773853078L;
+
+  private final TableId table;
+  private final List<RowToInsert> rows;
+  private final Boolean skipInvalidRows;
+  private final Boolean ignoreUnknownValues;
+
+  /**
+   * A Google Big Query row to be inserted into a table. Each {@code RowToInsert} has an associated
+   * id used by BigQuery to detect duplicate insertion requests on a best-effort basis.
+   *
+   * <p>
+   * Example usage of creating a row to insert:
+   * <pre>    {@code
+   *   List<Long> repeatedFieldValue = Arrays.asList(1L, 2L);
+   *   Map<String, Object> recordContent = new HashMap<String, Object>();
+   *   recordContent.put("subfieldName1", "value");
+   *   recordContent.put("subfieldName2", repeatedFieldValue);
+   *   Map<String, Object> rowContent = new HashMap<String, Object>();
+   *   rowContent.put("fieldName1", true);
+   *   rowContent.put("fieldName2", recordContent);
+   *   RowToInsert row = new RowToInsert("rowId", rowContent);
+   * }</pre>
+   *
+   * @see <a href ="https://cloud.google.com/bigquery/streaming-data-into-bigquery#dataconsistency">
+   *     Data Consistency</a>
+   */
+  public static class RowToInsert implements Serializable {
+
+    private static final long serialVersionUID = 8563060538219179157L;
+
+    private final String id;
+    private final Map<String, Object> content;
+
+    RowToInsert(String id, Map<String, Object> content) {
+      this.id = id;
+      this.content = content;
+    }
+
+    /**
+     * Returns the id associated with the row. Returns {@code null} if not set.
+     */
+    public String id() {
+      return id;
+    }
+
+    /**
+     * Returns the actual content of the row, as a map.
+     */
+    public Map<String, Object> content() {
+      return content;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("id", id)
+          .add("content", content)
+          .toString();
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(id, content);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (!(obj instanceof RowToInsert)) {
+        return false;
+      }
+      RowToInsert other = (RowToInsert) obj;
+      return Objects.equals(id, other.id)
+          && Objects.equals(content, other.content);
+    }
+
+    /**
+     * Creates a row to be inserted.
+     *
+     * @param id id of the row, used to identify duplicates
+     * @param content the actual content of the row
+     */
+    public static RowToInsert of(String id, Map<String, Object> content) {
+      return new RowToInsert(checkNotNull(id), checkNotNull(content));
```

Comment:
My idea was that if `RowToInsert` is used an `id` *must* be provided. If the user wants to insert a row without id `addRow(Map<String, Object> content)` can be used.
However, I don't mind adding a new factory method if you prefer.